library/src/main/java/com/timehop/stickyheadersrecyclerview/StickyRecyclerHeadersDecoration.java

```diff
@@ -101,9 +103,20 @@ public void onDrawOver(Canvas canvas, RecyclerView parent, RecyclerView.State st
       boolean hasStickyHeader = mHeaderPositionCalculator.hasStickyHeader(itemView, mOrientationProvider.getOrientation(parent), position);
       if (hasStickyHeader || mHeaderPositionCalculator.hasNewHeader(position, mOrientationProvider.isReverseLayout(parent))) {
         View header = mHeaderProvider.getHeader(parent, position);
-        Rect headerOffset = mHeaderPositionCalculator.getHeaderBounds(parent, header, itemView, hasStickyHeader);
+        //re-use existing Rect, if any.
+        Rect headerOffset = mHeaderRects.get(position);
+        if (headerOffset == null) {
+          headerOffset = new Rect();
+          mHeaderRects.put(position, headerOffset);
+        }
+        mHeaderPositionCalculator.initHeaderBounds(headerOffset, parent, header, itemView, hasStickyHeader);
         mRenderer.drawHeader(parent, canvas, header, headerOffset);
-        mHeaderRects.put(position, headerOffset);
+      }
+    }
+    //remove unused rects, if any.
+    if (childCount < mHeaderRects.size()) {
```

Comment:
I apologize for the slow response, and also if we're not entirely on the same page. You've put a lot more thought into this than I have, so please take my comments with a grain of salt and correct me if I'm wrong.

If items are removed from the list dynamically, I'm not sure this code would help. For example if you have 4 items in the adapter, and 3 of those are on screen (e.g. `getChildCount()`), and then you remove the 3rd one. `getChildCount()` will not change - there will still be 3 views attached to the RecyclerView, but the state of `mHeaderRects` will still be stale.

But I think in the big picture it doesn't matter if the values in `mHeaderRects` are stale or the incorrect size - the correct dimensions are calculated each time the view is drawn.  

As far as I can tell, this optimization really only saves memory at the cost of a few extra allocations.